zeroconf/__init__.py

```diff
@@ -1428,14 +1430,51 @@ def service_state_changed(self) -> SignalRegistrationInterface:
     def update_record(self, zc: 'Zeroconf', now: float, record: DNSRecord) -> None:
         """Callback invoked by Zeroconf when new information arrives.
 
-        Updates information required by browser in the Zeroconf cache."""
+        Updates information required by browser in the Zeroconf cache.
+
+        Ensures that there is are no unecessary duplicates in the list
+
+        """
 
         def enqueue_callback(state_change: ServiceStateChange, name: str) -> None:
-            self._handlers_to_call.append(
-                lambda zeroconf: self._service_state_changed.fire(
-                    zeroconf=zeroconf, service_type=self.type, name=name, state_change=state_change
-                )
-            )
+
```

Comment:
OK, two things here.

Regarding the logic, I had to write myself a little table to make sure I understand things correctly. I added some notes in two cases that I believe to be worth taking a second look at. The table:


| Existing enqueued state change | New state change to enqueue | State change on the list afterwards | My notes |
| ---------------------- | ----------------------------- | ------------------------------ | ---------- |
| |Added|Added|
| |Updated|Updated|
| |Removed|Removed|
|Added|Added|Added|
|Added|Updated|Added|
|Added|Removed|Added| Shouldn't this be just Removed, since the service is gone? When we enqueue Removed then there's Updated on the list already we switch to Removed.
|Updated|Added|Added|
|Updated|Updated|Updated|
|Updated|Removed|Removed|
|Removed|Added|Added|
|Removed|Updated|Removed| Now I'm not sure when would this happen, why wouldn't it make sense to result in Updated here?
|Removed|Removed|Removed|

Regarding the implementation details, I think `_handlers_to_call` could work better as an `OrderedDict[str, ServiceStateChange]` instead of `List[Tuple[str,ServiceStateChange]]`. The code below could then look more or less like (keeping the logic intact, I let myself reverse the logic and remove the early return(s), as it it wouldn't really make sense in this case):

```python
            if (
                (state_change is ServiceStateChange.Updated and name not in self._handlers_to_call)
                or (
                    state_change is ServiceStateChange.Removed
                    and self._handlers_to_call.get(name) is ServiceStateChange.Updated
                )
                or state_change is ServiceStateChange.Added
            ):
                self._handlers_to_call[name] = state_change
```

This removes all that `count()`, `remove()` and `append()` dance which is nice I think. Then the popping code would use [OrderedDict.popitem()](https://docs.python.org/3/library/collections.html#collections.OrderedDict.popitem) instead of `list.pop()`.

Reply:
Thinking about it, the code could just be an unconditional `self._handlers_to_call[name] = state_change`. :)

Reply:
I’ll have a bit more of a think about this one the coming week.

Reply:
Your last suggestion won't work because then an update trumps an add. I've used a slightly amended version of your first suggestion and the code you had above didn't update removed if for example the list was empty.

Regarding your 2 queries about the truth table, yes there are some combinations that don't make sense and in a well written responded shouldn't happen. However I've taken the view that I need to make a decision on the precedence of the messages and my view is that Added always wins, so for example if we receive an add and remove message in the same batch, then  add is the most likely meaning. I can hypothesise a reason this might occur if you have multiple res ponders and one goes offline. In this case Added + Removed still means there is a service there to respond. (note this is a contrived example as the messages would come in a different batch)

Reply:
Fair enough – if it causes issues I expect it'll be reported and we'll sort it out then, having a concrete case to deal with.